x/programs/rust/wasmlanche-sdk/src/memory.rs

```diff
@@ -52,9 +50,7 @@ impl From<HostPtr> for Vec<u8> {
         // drop will dealloc the bytes
         let host_ptr = ManuallyDrop::new(host_ptr);
 
-        let len = ALLOCATIONS
-            .with_borrow_mut(|allocations| allocations.remove(&host_ptr.0))
-            .expect("attempted to drop invalid host pointer");
+        let len = remove(host_ptr.0);
 
         unsafe { Vec::from_raw_parts(host_ptr.0.cast_mut(), len, len) }
```

Comment:
just checking my understanding -  `Vec::from_raw_parts` will take ownership of the `host_ptr`, meaning it will  automatically drop the Vec once its not longer in use. Can we do this without the first line
`let host_ptr = ManuallyDrop::new(host_ptr);`?

Reply:
So, when a `HostPtr` is dropped, it will remove the allocation data from the `ALLOCATIONS` map and attempt to dealloc the pointer by the length from the map. When we convert a `HostPtr` to a `Vec`, we don't want to perform the drop logic, we just want to drop the internal pointer (the stack value) and use the length from the `ALLOCATIONS` map. We use that data to construct the `Vec`. Now it's up to the `Vec` to dealloc on drop or move the data when it needs to grow, etc, etc. It's effectively like saying we're giving control over the bytes from `HostPtr` to `Vec`. 

That make sense?